### PR TITLE
Contributing additional Softmax Variants and refactored config to switch smoothly between them.

### PR DESCRIPTION
**Pull request by gkielian**
Added SigSoftmax (as nn.Module): https://arxiv.org/abs/1805.10829
Added a base 2 variation of SigSoftmax (as nn.Module)
Refactored Softermax into a nn.Module 

With each softmax variation inheriting from nn.Module, refactored to leverage polymorphism to simplify the selection of softmax variation.

Now simply use text setting within the Config, e.g. `softmax_variant = "softermax"` to choose or test out the softmax variation type.

### CODE REVIEW

File: train.py
Comment (klei22) [inline train.py:1]: Train.py should not have changed.

**Review by klei22 — CHANGES_REQUESTED** [review]
Probably should rebase, or at least remove changes to train.py

File: train.py
Comment (gkielian) [reply train.py:1]: Just did a rebase and updated in order to make sure train.py is up to date with main branch.

Looks like things are now good :+1:

**Comment by klei22** [comment]
Works except will need to remove the configurator,py mention in the sample.py script.

Should mention, noticed some `nans` when trying to train with SigSoftmax and the "Softermax" without xmax as False. 
We should keep a note of this somewhere.

**Comment by klei22** [comment]
Changes look good.

**Review by klei22 — APPROVED** [review]
